### PR TITLE
feat: annotate cli subcmd

### PR DESCRIPTION
**Pull request by atolopko-czi**
#### Reviewers
**Functional:** 

**Readability:** 
@MaximilianLombardo 

---

## Changes
- add
- remove
- modify

### CODE REVIEW

**Comment by atolopko-czi** [comment]
Initial stubbed annotate CLI subcommand help output, for discussion:

```
❯ cellxgene --help
Usage: cellxgene <options> COMMAND <args>

Options:
  -h, --help                      Show this message and exit.
  --version                       Show the software version and exit.
  --upgrade-check / --no-upgrade-check
                                  Check for release upgrades on start.  [default:
                                  upgrade-check]

Commands:
  annotate  Annotate H5AD file columns.Run `cellxgene annotation --help` for more
            information.
  launch    Launch the cellxgene data viewer. Run `cellxgene launch --help` for
            more information.
  prepare   Preprocess data for use with cellxgene. Run `cellxgene prepare --help`
            for more information.

cellxgene on  atolopko/2518-cell-type-prediction-cli-cmd [$] via 🐍 v3.9.12 (cellxgene-20220505) on ☁️
❯ cellxgene annotate --help
Usage: cellxgene annotate <options>

Options:
  -c, --annotation-column TEXT    The annotation column to be added.
  -a, --annotation-type [cell_type]
                                  The type of annotation to perform. This model to
                                  be used will be inferred from the annotation
                                  type.
  -h, --help                      Show this message and exit.
  -i, --input-h5ad-file TEXT      The input H5AD file containing the missing
                                  annotations.
  -o, --output-h5ad-file TEXT     The output H5AD file that will contain the
                                  generated annotation values. This option is
                                  mutually exclusive with --update-h5ad-file.
  -u, --update-h5ad-file          Flag indicating whether to update the input h5ad
                                  file with annotation values.  This option is
                                  mutually exclusive with --output-h5ad-file.
```

File: server/annotate/cell_type.py [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+    input_dataset = prepare_query_dataset(query_dataset, model)
+
+    # TODO: necessary to partition into AnnData-per-tissue, then invoke model on each partitioned AnnData object,
+    #  then reconstitute into a single AnnData
+    #
+    # TODO: record model metadata (version) in AnnData
+    predictions = model.predict(input_dataset)
+
+    query_dataset.obs[f"{annotation_column_name_prefix}"] = pd.Categorical(predictions)
+    # TODO: How do obtain confidence scores?
```
Comment (MaximilianLombardo) [inline server/annotate/cell_type.py:50]: @LuckyMD flagging this for your attention later

**Review by MaximilianLombardo — APPROVED** [review]
Thanks for putting this together Andrew - great start, so exciting 💯

File: server/annotate/cell_type.py [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+        model = pickle.load(f)
+        if not isinstance(model, SCANVI):
+            raise "invalid model object at {model_url}: expected type SCANVI, got {model.type}"
+        return model
+
+
+def annotate(query_dataset: AnnData, model: SCANVI, annotation_column_name_prefix: str) -> None:
+    input_dataset = prepare_query_dataset(query_dataset, model)
+
+    # TODO: necessary to partition into AnnData-per-tissue, then invoke model on each partitioned AnnData object,
```
[the remark this answers is not in the document]
Comment (atolopko-czi) [reply server/annotate/cell_type.py:43]: I was, but it sounded like we're just going with the design of "run one model on all cells, but allow multiple model runs to be performed".  So I'll eliminate this TODO.

File: server/annotate/cell_type.py [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+        var = ad.raw.var
+    else:
+        X = ad.X
+        var = ad.var
+    return AnnData(X=X, obs=ad.obs, var=var)
+
+
+# TODO: this should be part of the model's logic; note that build_raw_counts_adata() has been copied from the
+#  training pipeline code
+def prepare_query_dataset(query_dataset, model, raw_counts_input_layer_name=None) -> AnnData:
```
[the remark this answers is not in the document]
Comment (atolopko-czi) [reply server/annotate/cell_type.py:69]: If this is same as what `prepare --recipe` cli subcmd is doing, then maybe we keep it as a separate concern?  User does prediction step, then prepare, as needed?

File: server/annotate/cell_type.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (MaximilianLombardo) [reply server/annotate/cell_type.py:43]: That sounds good. Just to clarify, here's where I landed after yesterday's conversation:

- query dataset is always composed of a single tissue
- a reference data set could be composed of a single tissue or multiple tissues (we will focus on single tissue references - with one exception, the immune cell dataset from the teichmann lab)
- we will focus model training on the single tissue references and the immune multi-tissue reference for the MVP


Eliminating this TODO makes sense to me

File: server/annotate/cell_type.py (hunk 1 above)
Comment (LuckyMD) [reply server/annotate/cell_type.py:50]: Once I have some code, I will get to this.

Comment (MaximilianLombardo) [reply server/annotate/cell_type.py:57]: Thanks!

**Pull request closed without merging**